kolibri/core/content/api.py

```diff
@@ -686,6 +724,167 @@ def resume(self, request, **kwargs):
+    # 2. In the case where the tree topology means that this will not produce the full query, the limit of
+    #    25 immediate children and 25 grand children means that we are at most using 1 + 25 + 25 * 25 = 651
+    #    SQL parameters in the query to get the nodes for serialization - this means that we should not ever
+    #    run into an issue where we hit a SQL parameters limit in the queries in here.
+    # If we find that this page size is too high, we should lower it, but for the reasons noted above, we
+    # should not raise it.
+    page_size = 25
+
+    def consolidate(self, items, queryset):
+        if items:
+            return self.process_items(items, queryset)
+        return []
+
+    def validate_and_return_params(self, request):
+        depth = request.query_params.get("depth", 2)
+        lft__gt = request.query_params.get("lft__gt")
+
+        try:
+            depth = int(depth)
+            if 1 > depth or depth > 2:
+                raise ValueError
+        except ValueError:
+            raise ValidationError("Depth query parameter must have the value 1 or 2")
+
+        if lft__gt is not None:
+            try:
+                lft__gt = int(lft__gt)
+                if 1 > lft__gt:
+                    raise ValueError
+            except ValueError:
+                raise ValidationError(
+                    "lft__gt query parameter must be a positive integer if specified"
+                )
+        return depth, lft__gt
+
+    def get_grandchild_ids(self, child_ids, depth):
+        if depth == 2:
+            # Use this to keep track of how many grand children we have accumulated per child of the parent node
+            gc_by_parent = {}
+            # Iterate through the grand children of the parent node in lft order so we follow the tree traversal order
+            for gc in (
+                self.filter_queryset(self.get_queryset())
+                .filter(parent_id__in=child_ids)
+                .values("id", "parent_id")
+                .order_by("lft")
+            ):
+                # If we have not already added a list of nodes to the gc_by_parent map, initialize it here
+                if gc["parent_id"] not in gc_by_parent:
+                    gc_by_parent[gc["parent_id"]] = []
+                # If the number of grand children for a specific child node is less than the page size
+                # then we keep on adding them to both lists
+                # If not, we just skip this node, as we have already hit the page limit for the node that is
+                # its parent.
+                if len(gc_by_parent[gc["parent_id"]]) < self.page_size:
+                    gc_by_parent[gc["parent_id"]].append(gc["id"])
+                    yield gc["id"]
+
+    def retrieve(self, request, **kwargs):
+        """
+        A nested, paginated representation of the children and grandchildren of a specific node
+
+        :param request: request object
```

Comment:
shouldn't this doc add the request might have `depth` & `lft__gt`?

Reply:
Yes - I have been trying to look how best to document query params so they are picked up by our API explorer as well, will update this.